### ISSUE TITLE
ShadowResourcesTest intermittent failures

### ISSUE THREAD

**Issue by kriegfrj** [robolectric/robolectric#1795]
I am seeing the following failures on my machine:

````
Failed tests:
  ShadowResourcesTest.getThemeValueShouldSupportNotDereferencingResource:439 expected:<[1717044]4> but was:<[6656]4>
  ShadowResourcesTest.getThemeValueShouldSupportDereferenceResource:425 expected:<-1677[7216]> but was:<-1677[6961]>
````
A closer inspection of the second one revealed that this is a 4-byte color value which is being right-padded with 0xFF when it should be being right-padded with 0xFF. I haven't looked into the first yet but it looks like a slightly different issue (the hex for the two reported numbers is expecting:  0x0106000C, got 0x00010404). 

The failures are not consistent - sometimes one will fail, sometimes neither, sometimes both. Given that they don't seem to be an issue in Travis, I am suspecting one of two possibilities:

1. They happen only only on Windows.
2. They don't happen on a clean run - only on the second run.

I plan to investigate this but thought I would report the issue in case anyone else has seen this and has something useful to add that could help in tracking it down.

**Comment by kriegfrj** [comment]
Managed to figure out that this has something to do with static state being preserved between tests, so the intermittent nature of the failures is happening because of the randomised execution order. If I comment out all test methods other than these two then they seem to pass consistently. So it would seem that something in one of the other tests in `ShadowResourcesTest` is not cleaning up after itself properly and leaving the system in a state that causes these tests to fail.

Should now be a simple matter of narrowing down which of the other methods is causing the problem, and why. Usually when I've encountered this sort of problem before it's due to static state that isn't being cleared - I noticed that there is static state in `ShadowAssetManager` which isn't being reset so maybe that's the culprit. Further investigation will have to wait though.

**Comment by lexs** [comment]
Is there a way to somehow export the test order after a test run so you can rerun it in the same order?

**Comment by kriegfrj** [comment]
Thanks for the tip @lexs. Managed to use a JUnit rule based on `TestWatcher` to do exactly that and quickly pinpointed `applyStyleForced()` as the culprit. It seems that the applied styles are being remembered between tests because the internal theme IDs get reused between tests. This is affecting the outcome of the two test methods in question.

I tried implementing a static resetter method for `ShadowAssetManager` which clears `APPLIED_STYLES`, and that seems to fix this issue. However, I am having difficulty coming up with a test that directly exercises this bug in a way that is reliable & repeatable, which ideally I'd want to include too as part of a TDD methodology and to help prevent regressions. Suggestions welcome.

**Comment by lexs** [comment]
@kriegfrj Will resetting `APPLIED_STYLES ` affect performance?

**Comment by kriegfrj** [comment]
Agreed, it shouldn't make a significant difference to performance. At any rate, while speed is important, stability and accuracy are probably more important.

**Comment by kriegfrj** [comment]
Submitted a fix for this (#1808) - instead of a resetter, I thought it would be better to make the applied styles state non-static so it automatically gets cleaned up at the end of each test when the `AssetManager` itself gets cleaned up,
